### ISSUE TITLE
--OjaNewton numerical instability and slowness

### ISSUE THREAD

**Issue by arielf** [JohnLangford/vowpal_wabbit#1017]
Tried to use the new `--OjaNewton` (2nd order learning) on `test/train-sets/0002.dat`

Observations:

1) With vanilla run (all defaults) results are surprisingly much worse than without it (final mean loss is 0.127483 vs 0.003847.  About ~30x greater).  Note that this data-set labels are all around 0.5.  Predictions with `--OjaNewton` are way off, e.g. rather than roughly converge on the 0.5 average, many (not all) of the examples cause `--OjaNewton` to predict either 0.0 or 1.0 and the mean loss to jump sharply upwards.

2) Adding `-b 21` (or any greater value) makes `--OjaNewton` run many times slower (`-b 24` takes half a minute to complete one pass, about 100 times slower). At `-b 21` I see NaNs on all examples and final average loss more than doubling to 0.283062.  This is weird, why would changing `-b` affect the convergence?

**Comment by JohnLangford** [comment]
Haipeng, can you comment on this?

-John

**Comment by haipengl** [comment]
Sorry for the delay. It seems like the problem is in the learning rate. If you run OjaNewton with --alpha_inverse=0.1, you will get similar loss, although indeed it doesn't seem like second order information is helping here...

**Comment by arielf** [comment]
Should we make `--alpha_inverse=0.1` the default then?

One of the nice features of `vw` is that defaults work well so tweaking of secondary parameters isn't necessary in many cases.

**Comment by haipengl** [comment]
I set the current default based on my experience of running this algorithm on a bunch of datasets (basically those in the paper). Do you find using --alpha_inverse=0.1 is in general better?

**Comment by arielf** [comment]
@haipengl not generally better, but it bothers me that my first attempt to use the `--OjaNewton` option with nothing else (vanilla run) resulted in such unexpected outcome.  Not sure what the best solution is. The default learning rate in `vw` is, I believe 0.5 and it works pretty well on most of the data-sets in our test suite.   Perhaps this should be documented in the wiki: `--OjaNewton` performance is particulaily sensitive to its learning rate which can be specified via the new option: --alpha_inverse=<value>` .
What do you think?

**Comment by arielf** [comment]
Also: the `-b` dependence seems like a bug that is unrelated to convergence and learning rates.

For example.   The following two runs produce vastly different results. The only difference is `-b 20` vs `-b 22`:

```
$ vw-hypersearch 0.0001 0.5 vowpalwabbit/vw -b 20 --OjaNewton --alpha_inverse=% test/train-sets/0002.dat
trying 0.309055190976072 .......... 0.032052 (best)
trying 0.191044809023928 .......... 0.009237 (best)
trying 0.118110381952145 .......... 0.004499 (best)
trying 0.0730344270717827 .......... 0.004497 (best)
trying 0.0451759548803622 .......... 0.011763
trying 0.0902519097607243 .......... 0.004002 (best)
trying 0.100892899263203 .......... 0.004067
trying 0.0836754165742616 .......... 0.004076
trying 0.0943164060767404 .......... 0.004003
trying 0.0877399128902777 .......... 0.004018
trying 0.0918044092062938 .......... 0.003999 (best)
trying 0.092763906631171 .......... 0.003999 (best)
loss(0.0927639) == loss(0.0918044): 0.003999
trying 0.0922841579187324 .......... 0.003999 (best)
0.0922842       0.003999

$ vw-hypersearch 0.5 0.0001 0.5 vowpalwabbit/vw -b 22 --OjaNewton --alpha_inverse=% test/train-sets/0002.dat
trying 0.309055190976072 .......... 0.283062 (best)
trying 0.191044809023928 .......... 0.283062 (best)
loss(0.309055) == loss(0.191045): 0.283062
trying 0.25005 .......... 0.283062 (best)
0.25005 0.283062
```

**Comment by haipengl** [comment]
Adding a note in the wiki is definitely a good idea (I am new to Github and not sure how to do it though...)

What does -b do exactly?

**Comment by JohnLangford** [comment]
-b changes the number of bits in the hash function (and hence, the size of the model).  You should have an option to edit things in the wiki https://github.com/JohnLangford/vowpal_wabbit/wiki

**Comment by haipengl** [comment]
I see. Will work on the wiki later then.

I really don't understand why -b changes things then. Will take a look into it...

**Comment by JohnLangford** [comment]
When there are collisions, parameters become shared which causes a loss of expressivity.  

Have you edited the wiki? I'd like to close this.

**Comment by haipengl** [comment]
Yes, I just did.

Haipeng
